### PR TITLE
Desktop - Invity Buy - Change popup flow to default browser

### LINKED ISSUE #2711: Desktop - Invity Buy - Change popup flow to default browser
As discussed during the review, the popup flow should be replaced with a buy flow using the default browser and a local web server to catch incoming information (see oauth example).

### LINKED ISSUE #2757: Invity - Buy - flow blocked with any provider
**Info:**
 - OS: macOS Catalina 10.15.7
 - Suite: app 20.11.0 staging 34e505ec96de077e9064e76ac4e835e62d53a338
 - Device: model T 2.3.4 50854b921
 - API server: staging-exchange
 
**Describe the bug**
Buy flow cannot be finished in the app version of Suite - provider's payment gateway.

**Steps to reproduce:**
1. Launch Suite and navigate to Accounts -> Trade -> Buy,
2. Enter an arbitrary amount above minimum and confirm,
3. Select any offer (tested Simplex) and confirm,
4. Confirm receiving address on device and select The Finish transaction button.

**Actual result:**
A new electron app window is opened with only the header displayed; no content on the page is displayed. Same applies to the other providers (mercuryo, BTC Direct).

**Expected result:**

# Note: None of the testing devices have VPNs installed.

**Screenshots:**
**Simplex**
image
**mercuryo**
image
**BTC Direct**
image

### PR DESCRIPTION
**Pull request by martinboehm**
This pull request implement opening of buy crypto payment gate windows in a default browser window (instead of using windows inside electron app) . The PR is quite big, because it addresses the following issues:
- opening payment gate using GET method
- opening payment gate using POST method as required by Simplex - it requires usage of the internal server to serve a page with a form, which is then submitted to Simplex
- go to buy status detail page in the Suite after opening the payment gate
- return to the Suite from the payment gate using internal server, with logic which skips unnecessary redirect when the user is already on buy detail status page
- fix a bug in reducer (backport from branch in development feat/invity-exchange), which caused incorrect refresh of previous transactions - the bug did not exhibit itself in the web version as the application is fully reloaded on return from the payment gate

This PR fixes #2711 and fixes https://github.com/trezor/trezor-suite/issues/2757.

Comment: in general, all external urls (payment gate servers and links to support/status pages of the partners) are dynamically obtained from our partners and are not known up front. The fixed list of allowedExternalUrls will therefore cause problems to Invity functionality. I think we should discuss the issue.

### CODE REVIEW

**Comment by martinboehm** [comment]
Rebased due to conflict with develop.

File: packages/suite-desktop/src-electron/http-receiver.ts [hunk 1]
```diff
@@ -139,6 +148,41 @@ export class HttpReceiver extends EventEmitter {
...
+        response.end(template);
+    };
+
+    private buyPostSubmitHandler = (request: Request, response: http.ServerResponse) => {
+        const { query } = url.parse(request.url, true);
+        const action = query.a;
+        const template = `
+         <body>
+            Forwarding to ${action}...
+            <form id="buy-form" method="POST" action="${action}">
```
Comment (szymonlesisz) [inline packages/suite-desktop/src-electron/http-receiver.ts:172]: POST request can be done in pure javascript without html
```
<script type="text/javascript">
    var xhr = new XMLHttpRequest();
    xhr.open("POST", query.a, true);
    xhr.send(query); 
    // or xhr.setRequestHeader('Content-Type', 'application/json'); xhr.send(JSON.stringify(query)); not sure which format is accepted
</script>
```

pros:
you can handle request errors using `xhr.onreadystatechange`

File: packages/suite-desktop/src-electron/electron.ts [hunk 2]
```diff
@@ -371,20 +366,43 @@ httpReceiver.on('server/listening', () => {
         app.focus();
     });
 
+    httpReceiver.on('buy/redirect', url => {
+        // if we are already in the buy detail page of the correct account, we do not have to redirect
```
Comment (szymonlesisz) [inline packages/suite-desktop/src-electron/electron.ts]: pls move this code to a separate file (suite-desktop/src-electron/buy.ts ?)
it doesn't look good here and it's only temporary (as discussed in person)

**Comment by martinboehm** [comment]
Rebased after another conflict.

File: packages/suite/src/hooks/wallet/useCoinmarketBuyOffers.ts [hunk 3]
```diff
@@ -15,6 +15,7 @@ import { Props, ContextValues } from '@wallet-types/coinmarketBuyOffers';
 import { useSelector } from 'react-redux';
```
Comment (szymonlesisz) [inline packages/suite/src/hooks/wallet/useCoinmarketBuyOffers.ts]: { useSelector} from '@suite-hooks';

- its already typed
- it's memoized (default one is not, this leads to unnecessary render on EVERY store change) fixed here: https://github.com/trezor/trezor-suite/commit/a3061e0a1ceaf397b44200cfae7405fcb819b104

Comment (szymonlesisz) [reply packages/suite/src/hooks/wallet/useCoinmarketBuyOffers.ts]: i've noticed that `useSelector` from 'react-redux' is used in few other places in suite, i'll address that in new PR

File: packages/suite/src/hooks/wallet/useCoinmarketBuyOffers.ts [hunk 4]
```diff
@@ -147,10 +150,17 @@ export const useOffers = (props: Props) => {
             });
         } else {
             await saveTrade(response.trade, account, new Date().toISOString());
-            // eslint-disable-next-line no-lonely-if
             if (response.tradeForm) {
                 submitRequestForm(response.tradeForm);
             }
+            if (isDesktop()) {
+                await saveTransactionDetailId(response.trade.paymentId);
+                goto('wallet-coinmarket-buy-detail', {
```
Comment (szymonlesisz) [inline packages/suite/src/hooks/wallet/useCoinmarketBuyOffers.ts]: `goto('wallet-coinmarket-buy-detail', selectedAccount.params)`

**Review by szymonlesisz — CHANGES_REQUESTED** [review]

**Comment by martinboehm** [comment]
The issues apart from the POST handler are resolved.

File: packages/suite/src/utils/wallet/coinmarket/buyUtils.ts [hunk 5]
```diff
@@ -126,31 +129,30 @@ function addHiddenFieldToForm(form: any, fieldName: string, fieldValue: any) {
...
-        const formWindow = window.open('', invityWindowName);
-        if (formWindow) {
-            formWindow.document.body.appendChild(form);
-            form.submit();
-        }
+        let params = `a=${encodeURIComponent(tradeForm.form.formAction)}`;
+        Object.keys(fields).forEach(k => {
+            params += `&${k}=${encodeURIComponent(fields[k])}`;
+        });
+        window.open(`http://127.0.0.1:21335/buy-post?${params}`, '_blank');
```
Comment (szymonlesisz) [inline packages/suite/src/utils/wallet/coinmarket/buyUtils.ts]: use ELECTRON_RECEIVER_SERVER here

**Review by szymonlesisz — CHANGES_REQUESTED** [review]

File: packages/suite/src/hooks/wallet/useCoinmarketBuyOffers.ts [hunk 6]
```diff
@@ -43,17 +43,18 @@ export const useOffers = (props: Props) => {
...
         openCoinmarketBuyConfirmModal: coinmarketBuyActions.openCoinmarketBuyConfirmModal,
         addNotification: notificationActions.addToast,
+        saveTransactionDetailId: coinmarketBuyActions.saveTransactionDetailId,
     });
 
-    const invityAPIUrl = useSelector<
-        AppState,
-        AppState['suite']['settings']['debug']['invityAPIUrl']
-    >(state => state.suite.settings.debug.invityAPIUrl);
+    const invityAPIUrl = useSelector<AppState['suite']['settings']['debug']['invityAPIUrl']>(
```
Comment (szymonlesisz) [inline packages/suite/src/hooks/wallet/useCoinmarketBuyOffers.ts]: you don need to use type anymore, it should be typed already:
`const invityAPIUrl = useSelector(state => state.suite.settings.debug.invityAPIUrl);`

**Comment by martinboehm** [comment]
Done

**Review by szymonlesisz — APPROVED** [review]
conditionally approved regarding to commit messages (will be squashed)
Electron part `handleExternalUrls` and `buyUtils` should be refactored as discussed in person

**Pull request merged**